### PR TITLE
Experimental Implementation of the WindowsAppRuntime Deployment API

### PR DESCRIPTION
**Pull request by dkbennett**
Experimental full implementation of the WindowsAppRuntime Deployment API, described in #1253

The IDL is marked [Experimental] and the implementation and tests are gated behind a TerminalVelocity Feature.

### CODE REVIEW

File: dev/Deployment/DeploymentManager.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            return status;
+        }
+
+        std::wstring frameworkPackageFullName{ packageFullName };
+        auto initializeResult{ DeployPackages(frameworkPackageFullName) };
+        auto initializeStatus{ winrt::make<implementation::DeploymentStatus>(SUCCEEDED(initializeResult), FAILED(initializeResult), initializeResult) };
+        return initializeStatus;
+    }
+
+    MddCore::PackageInfo DeploymentManager::GetPackageInfoForPackage(std::wstring const& packageFullName)
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentManager.cpp:88]: `MddCore::*` weren't expected to be an API (even an internal one). But that's cool. I'll work on polishing/liberating these things in the future so it's more generally useful (cleanly, w/o unexpected future surprises :-)

File: dev/Deployment/DeploymentManager.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            const auto packageFullName{ packageFullNames[index] };
+            packageFullNamesList.push_back(std::wstring(packageFullName));
+        }
+        return packageFullNamesList;
+    }
+
+    HRESULT DeploymentManager::VerifyPackage(const std::wstring& packageFamilyName, const PACKAGE_VERSION targetVersion) try
+    {
+        auto packageFullNames{ FindPackagesByFamily(packageFamilyName) };
+        bool match{ false };
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentManager.cpp:127]: NIT: `bool match{};`

File: dev/Deployment/DeploymentManager.cpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        for (const auto& packageFullName : packageFullNames)
+        {
+            auto packagePath{ GetPackagePath(packageFullName) };
+            if (packagePath.length() == 0)
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentManager.cpp:131]: `if (packagePath.empty())` ?

File: dev/Deployment/DeploymentManager.cpp [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            if (packageId.Version().Version >= targetVersion.Version)
+            {
+                match = true;
+                break;
+            }
+        }
+
+        if (!match)
+        {
+            return HRESULT_FROM_WIN32(ERROR_NOT_FOUND);
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentManager.cpp:146]: Would you want logging here? If so `RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !match);` ?

File: dev/Deployment/DeploymentStatus.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        DeploymentStatus() = default;
+
+        DeploymentStatus(bool isOK, bool packageInstallRequired, winrt::hresult const& extendedError);
+        bool IsOK();
+        bool PackageInstallRequired();
+        winrt::hresult ExtendedError();
+
+    private:
+        const bool m_isOK{ false };
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentStatus.h:18]: NIT: `...{};`

**Review by DrusTheAxe — APPROVED** [review]
One comment of note for a future PR

File: dev/Deployment/DeploymentManager.cpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+        // Should only be called with a framework name that exists.
+        FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), frameworkPackageInfo.Count() != 1);
+
+        // Only the Microsoft Publisher Id is supported.
+        auto expectedPublisherId{ frameworkPackageInfo.Package(0).packageId.publisherId };
+        FAIL_FAST_HR_IF(E_INVALIDARG, CompareStringOrdinal(expectedPublisherId, -1, WINDOWSAPPRUNTIME_IDENTITY_PUBLISHERID, -1, TRUE) != CSTR_EQUAL);
+
+        // The framework naming scheme consists of a prefix (ex: "Microsoft.WindowsAppSDK.") and a possible
+        // postfix (ex: "-1.0-Release"). We assume all packages of this release match this similar naming scheme
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply dev/Deployment/DeploymentManager.cpp]: Fixed

File: dev/Deployment/DeploymentManager.cpp [hunk 7]
```diff
@@ -29,25 +29,30 @@ namespace winrt::Microsoft::Windows::ApplicationModel::WindowsAppSDK::implementa
...
-        std::wstring frameworkFamilyName{ frameworkPackageInfo.Package(0).packageFamilyName };
-        auto posFrameworkIdentifier{ frameworkFamilyName.find(WINDOWSAPPRUNTIME_FRAMEWORK_PACKAGE_IDENTIFIER) };
-        const int c_frameworkIdentifierLength{ ARRAYSIZE(WINDOWSAPPRUNTIME_FRAMEWORK_PACKAGE_IDENTIFIER) - 1 };
-        auto posFrameworkDelimeter{ frameworkFamilyName.find(WINDOWSAPPRUNTIME_NAME_DELIMETER) };
-        auto packageFamilyPrefix{ frameworkFamilyName.substr(0, posFrameworkIdentifier)};
-        auto packageFamilyPostfix{ frameworkFamilyName.substr(posFrameworkIdentifier + c_frameworkIdentifierLength, posFrameworkDelimeter - (posFrameworkIdentifier + c_frameworkIdentifierLength))};
+        FAIL_FAST_HR_IF(E_INVALIDARG, CompareStringOrdinal(expectedPublisherId, -1, WINDOWSAPPRUNTIME_PACKAGE_PUBLISHERID, -1, TRUE) != CSTR_EQUAL);
+
+        // The framework naming scheme consists of a prefix (ex: "Microsoft.WindowsAppRuntime.")
+        // and a possible VersionTag (ex: "-1.0-Release"). We assume all packages of this release
```
Comment (DrusTheAxe) [inline dev/Deployment/DeploymentManager.cpp]: NIT: "ex: 1.0-Preview2"

There is no "Release" versiontag. We use channel names as versiontag (with "" for Stable channel).

And we use dot (not dash) between the name and release (1.0 = major.minor= release, or release version, to be pedantic). So technically the versiontag is "preview2", the release version is "1.0", we use a dash delimiter between them and we use a dot delimiter between the name and the version.

**Review by DrusTheAxe — APPROVED** [review]

File: dev/Deployment/DeploymentManager.cpp (hunk 7 above)
Comment (dkbennett) [reply dev/Deployment/DeploymentManager.cpp]: Fixed in an earlier change

**Pull request merged**